channels/offline-channel-js/src/OfflineChannel.ts

```diff
@@ -0,0 +1,604 @@
+                        // Just handling any rejection to avoid an unhandled rejection event
+                    });
+                }
+                
+            }
+
+            function _setRetryTime(linearFactor: number = 1) {
+                const SlotDelayInSeconds = 10;
+                let delayInSeconds: number;
+        
+                if (_consecutiveErrors <= 1) {
+                    delayInSeconds = SlotDelayInSeconds;
+                } else {
+                    const backOffSlot = (Math.pow(2, _consecutiveErrors) - 1) / 2;
+                    // tslint:disable-next-line:insecure-random
+                    let backOffDelay = Math.floor(Math.random() * backOffSlot * SlotDelayInSeconds) + 1;
+                    backOffDelay = linearFactor * backOffDelay;
+                    delayInSeconds = Math.max(Math.min(backOffDelay, 3600), SlotDelayInSeconds);
+                }
+        
+                // TODO: Log the backoff time like the C# version does.
+                let retryAfterTimeSpan = dateNow() + (delayInSeconds * 1000);
+        
+                // TODO: Log the retry at time like the C# version does.
+                _retryAt = retryAfterTimeSpan;
+            }
+
+            function _getCoreItemCtx(coreConfig: IConfiguration & IConfig, core: IAppInsightsCore, extensions: IPlugin[], pluginChain?: ITelemetryPluginChain) {
+                if (coreConfig) {
+                    // Make sure the extensionConfig exists
+                    coreConfig.extensionConfig = coreConfig.extensionConfig || {};
+                }
+
+                if (!pluginChain && core) {
+                    // Get the first plugin from the core
+                    pluginChain = core.getProcessTelContext().getNext();
+                }
+
+                let nextPlugin: IPlugin = null;
+                let rootNext = _self._getTelCtx().getNext();
+                if (rootNext) {
+                    nextPlugin = rootNext.getPlugin();
+                }
+
+                return createProcessTelemetryContext(pluginChain, coreConfig, core, nextPlugin);
+            }
+
+
+            function _getPayload(evt: IPostTransmissionTelemetryItem | ITelemetryItem) {
+                try {
+
+                    if (evt) {
+                        evt.iKey = evt.iKey || _urlCfg.iKey;
+                        return _offineSupport.serialize(evt, _convertUndefined);
+                    }
+                    
+                } catch (e) {
+                    // eslint-disable-next-line no-empty
+                }
+                return null;
+                
+            }
+
+            function _constructPayloadData(payloadArr: string[], criticalCnt?: number) {
+
+                try {
+                    if (!_offineSupport) {
+                        return null;
+                    }
+                    let cnt = criticalCnt || 0;
+                    let payload = _offineSupport.batch(payloadArr);
```

Comment:
We should probably change this to return the IPayloadData as the PostChannel / HttpManager may be configured to append additional values to the query string and / or create headers from them -- so we don't want to duplicate that code here as Sender doesn't have that level of complexity.